### PR TITLE
WIP: Update maintainers guide

### PR DESCRIPTION
**Pull request by tux-rampage**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | yes?
| New Feature   | no
| RFC           | yes
| QA            | yes

### Description

As discussed with @weierophinney and the contributors channel, the maintainers guide needs to be updated to the automatic releases workflow.

This PR is just a starting point to drive this document with the technical steering committee to a guide that helps maintainers to maintain the component as intended and have a point of orientation to refer to.

I just incorporated the new process as much as I understood it - which might be off or even totally wrong. So this might also help me and potential other maintainers, who are not part of the steering committee, to better understand the intended process.

### CODE REVIEW

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:31]: This part isn't entirely true or necessary. There can be a single branch with no releases existing (e.g., a 0.1.x branch), or where the migration encountered a `develop` branch that had no changes from the `master` branch, and we only published the release branch associated with the most recent release.

AFTER a release has been made using the automatic-releases workflow, there will always be a minimum of two release branches. But BEFORE, there may only be one.

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:33]:
```suggestion
- only unreleased branches may receive *new features*.
```

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:41]: This part changes.

We don't have LTS anymore.

We have "security windows". Essentially, in the case of a security advisory and associated release(s), we make the patch to the earliest release branch we still support. From there, automatic-releases takes over, and does merge-up releases to each next latest release branch until it hits the latest. Each of those releases along the way will be listed in the security advisory.

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:126]: This workflow works pretty much the opposite now.

Since the default branch is the latest release branch, the more common scenario is that somebody will submit a bugfix against the latest release branch, when it should go against the previous. In those cases, you'll need to re-target the PR and help the user rebase or cherry-pick against the previous release branch.

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:130]: I think we only want to target the lowest affected release branch for _security fixes_. For bugfixes, the latest release branch with a release created from it would be enough.

So, as an example, let's say we had a repo with the following release branches:

- 2.7.x
- 2.8.x
- 2.9.x

and tags:

- 2.7.0
- 2.7.1
- 2.7.2
- 2.8.0
- 2.8.1

We'd apply a bugfix against the 2.8.x branch in that case, and then end up issuing a 2.8.2 release.

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:131]:
```suggestion
- If the PR is a new feature, merge to the latest existing release branch against which no tags have been created.
  If none exists, create a new one from the latest existing release branch using `git switch -c {X.Y+1.x} {X.Y.x}`.
```

**Review by weierophinney — CHANGES_REQUESTED** [review]
Awesome start, thank you!

I've left a few notes, as well as a few code suggestions.

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:26]:
```suggestion
We use [automatic releases](https://github.com/laminas/automatic-releases) for
```

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:27]: Should be "Laminas Project", not "laminas Framework".

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:143]: I don't think we we ever discussed this in the TSC. What's the rationale for "create a merge commit" as the selected option? "Rebase and merge" leads to a more linear tree, which makes granular reversions of changes easier, while "Squash and merge" makes it easy to back out the entire changeset at once (though I'd argue squash and merge is _never_ a good idea because you lose granularlity). I've often used the `merge --ff` option when manually merging, which is similar to "Rebase and merge".

Maybe @Ocramius could weigh in?

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:146]: This doesn't detail _how_ to merge. Should it be a standard merge, with a merge commit? Or a `merge --ff`? Or should we rebase and then `merge --ff`? (This discussion is related to my comments about using the button for merging. We need consistency, but also a good rationale for what we choose. And it needs to be the same whether using the button, or doing it at the CLI.)

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:185]:
```suggestion
noted; things like coding standards fixes, continuous integration changes, or typo fixes do not need
to be communicated. However, any addition, deprecation, removal, or fix with user impact
SHOULD be noted. Please provide a succinct but *narrative* description for the change you merge,
detailing the change, its impact on the user, and what changes they may need to make to their code.
```

File: MAINTAINERS.md
Comment (weierophinney) [inline MAINTAINERS.md:191]:
```suggestion
Releases are automatically performed by closing a milestone. All issues and pull
requests in the milestone MUST be closed before closing the milestone; otherwise,
the release will not be issued.
```

**Review by weierophinney — CHANGES_REQUESTED** [review]
A few more changes, and a couple points for discussion.

File: MAINTAINERS.md
Comment (tux-rampage) [reply MAINTAINERS.md:143]: Actually this is just a suggestion since there is official statement here and I tried to base this on the previous gitflow strategy. But that all is a matter of taste and whatever is deemed to be important about the commit history. 

- A merge commit makes the origin of a change more obvious in the commit graph IMHO. 
- A squashed commit will consider the PR as a single commit which is sometimes useful to get rid of small adjustment commits during review rounds. It requires discipline to create small changesets though (which is always encouraged from a maintainers pov) - but this might drive off some contributors that are not familiar with this process.
- With a rebase and merge you'd loose the ability to track commits for its origin, but it produces a linear commit history.

I could also rephrase it to a "Pick whatever suits the individual situation for the PR" - which might be the best option here.

File: MAINTAINERS.md
Comment (tux-rampage) [reply MAINTAINERS.md:146]: That's right much like above I actually don't know how to start here. At first do we want a permissive approach or restrictive approach. Both has its pros and cons. Permissive offers more flexibility, restrictive is more streamlined. Whatever the descition is: Here it should be clear to a Maintainer, what's good to go and what should not be done in either a permissive ("Don't do x, y, z - otherwise do as you think it suits best") or restrictive ("Only do x, y, z") way.

**Comment by tux-rampage** [comment]
@weierophinney thanks for taking time to review - I'll pull in your changes, but not sure when I'll be able to do so. I'm quite packed with contractor work atm.

**Comment by weierophinney** [comment]
@tux-rampage I've pushed changes based on feedback I provided, some feedback I obtained from this week's TSC meeting, and a complete review of the document (I was able to greatly simplify the release and changelog sections). I've also moved to one-sentence-per-line structure, which is what we're aiming for in the documentation going forwards.

I'm going to open this for review, and ask other TSC members to take a look.

**Review by Ocramius — APPROVED** [review]
Skimmed, and I think overall it makes sense. Especially like the new nomination template idea :+1:

**Review by weierophinney — APPROVED** [review]

**Pull request merged**
